Fix cmake configuration when lto flags not available

jagerman opened the pull request:
CMake current fails to run with Clang on linux:

    $ CC=clang CXX=clang++ cmake ..
    ... (various configuration) ...
    -- Configuring done
    CMake Error at tools/pybind11Tools.cmake:135 (target_compile_options):
      Error evaluating generator expression:

        $<:-flto>

      Expression did not evaluate to a known generator expression
    Call Stack (most recent call first):
      tests/CMakeLists.txt:68 (pybind11_add_module)

This commit fixes the generator expressions when the lto flags are not found (they will be empty strings, but the generator expression needs a 0).

jagerman commented:
Actually, this is a little deeper problem.  `-flto` *is* supported, but, at least with clang on linux, it has to be passed to both the compiler *and* the linker.  `CHECK_CXX_COMPILER_FLAG` only specifies it when compiling; when it invokes clang again to link *without* the `-flto` option, clang produces an unknown file type error, and so cmake aborts, returning that the flag isn't supported.

I don't see an obvious way in cmake to do this, so I'll hack something into the pybind lto-determining function to also (temporarily) set the linker flag when testing.

jagerman commented:
Fixed now to also pass the flag to the linker when testing, and to add it to the target linker options when returning; `-flto` gets added correctly for clang on linux now.

(The first patch is still needed for cases when `-flto` isn't supported).

wjakob commented:
Here is an alternative design which I use in my own projects. It does a few more things that may be helpful to integrate here.

```cmake
if (U_CMAKE_BUILD_TYPE MATCHES REL AND CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Clang)$")
  # Enable link time optimization
  if (NOT CMAKE_CXX_FLAGS MATCHES "-flto")
    if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
      set(LTO_CXX_FLAGS "-flto=thin")
      set(LTO_LINKER_FLAGS "-flto=thin")
      if (NOT APPLE AND U_CMAKE_BUILD_TYPE MATCHES MINSIZEREL)
          # Clang Gold plugin does not support -Os
          set(LTO_CXX_FLAGS "${LTO_CXX_FLAGS} -O3")
      endif()
    else()
      set(LTO_CXX_FLAGS "-flto -fno-fat-lto-objects")
      set(LTO_LINKER_FLAGS "-flto")
    endif()

    CHECK_CXX_COMPILER_AND_LINKER_FLAGS(HAS_LTO ${LTO_CXX_FLAGS} ${LTO_LINKER_FLAGS})

    if (HAS_LTO)
      message(STATUS "LTO support enabled.")
      set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${LTO_LINKER_FLAGS}")
      set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${LTO_LINKER_FLAGS}")
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LTO_CXX_FLAGS}")
      set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${LTO_CXX_FLAGS}")
    else()
      message(STATUS "LTO not supported by the compiler.")
    endif()
  endif()
elseif(MSVC)
  set(Configurations RELEASE RELWITHDEBINFO MINSIZEREL)
  set(LinkTypes EXE SHARED MODULE STATIC)
  foreach(Configuration ${Configurations})
    set("CMAKE_CXX_FLAGS_${Configuration}" "${CMAKE_CXX_FLAGS_${Configuration}} /GL")
    foreach(LinkType ${LinkTypes})
      set("CMAKE_${LinkType}_LINKER_FLAGS_${Configuration}" "${CMAKE_${LinkType}_LINKER_FLAGS_${Configuration}} /LTCG")
    endforeach()
  endforeach()
  message(STATUS "LTO support enabled.")
endif()
```

The macro ``CHECK_CXX_COMPILER_AND_LINKER_FLAGS`` is defined as

```cmake
include(CheckCXXSourceRuns)
macro(CHECK_CXX_COMPILER_AND_LINKER_FLAGS _RESULT _CXX_FLAGS _LINKER_FLAGS)
  set(CMAKE_REQUIRED_FLAGS ${_CXX_FLAGS})
  set(CMAKE_REQUIRED_LIBRARIES ${_LINKER_FLAGS})
  set(CMAKE_REQUIRED_QUIET TRUE)
  check_cxx_source_runs("int main(int argc, char **argv) { return 0; }" ${_RESULT})
  set(CMAKE_REQUIRED_FLAGS "")
  set(CMAKE_REQUIRED_LIBRARIES "")
endmacro()
```

wjakob commented:
.. and in yet another project, the path for Clang is a bit more advanced: I've found it necessary to inform Gold/Darwin linker where they should store their caches for thin LTO to actually kick in.

```cmake
file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/.cache)
if (APPLE)
  set(LTO_LINKER_FLAGS    "-flto=thin -Wl,-cache_path_lto,${CMAKE_CURRENT_BINARY_DIR}/.cache")
else()
  set(LTO_LINKER_FLAGS    "-flto=thin -Wl,-plugin-opt,cache-dir=${CMAKE_CURRENT_BINARY_DIR}/.cache")
endif()
```

wjakob commented:
(These may or may not apply. I thought it would be useful to assemble all the collected information about enforcing LTO/thin LTO on various platforms)

wjakob commented:
Note also the ``-fno-fat-lto-objects`` passed to GCC, which is quite important to avoid running the compiler backend twice.

jagerman commented on tools/pybind11Tools.cmake:
```diff
@@ -37,25 +37,31 @@ function(select_cxx_standard)
+  set(orig_CMAKE_REQUIRED_LIBRARIES "${CMAKE_REQUIRED_LIBRARIES}")
+  foreach(f ${try_flags})
+    # clang, in particular, needs -flto to be provided at link time as well; it shouldn't hurt the others
+    set(CMAKE_REQUIRED_LIBRARIES "${orig_CMAKE_REQUIRED_LIBRARIES} ${f}")
```
I changed it from a macro to a function--the main difference being that cmake does an implicit push/pop of variables, so the sets here are local (the exception being the PARENT_SCOPE one later on).

jagerman commented:
That seems like a better implementation than mine.  I'll update the PR with it shortly.

jagerman commented:
That's incomplete, still working on it.

jagerman commented:
I think this is working now.

rdeits commented:
I can confirm that this fixes the issue I reported in #639

rdeits commented on tools/pybind11Tools.cmake:
```diff
@@ -37,25 +37,74 @@ function(select_cxx_standard)
+  set(try_cxx_flags "")
+  set(try_linker_flags "")
+
+  if(U_CMAKE_BUILD_TYPE MATCHES REL AND CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Clang)$")
```
This check doesn't match the compiler on OSX, which is "AppleClang". Is that intended?

jagerman replied:
I wonder if that was why @wjakob needed the ".. and in yet another project" workaround.

rdeits commented on tools/pybind11Tools.cmake:
```diff
@@ -37,25 +37,74 @@ function(select_cxx_standard)
+  set(CMAKE_REQUIRED_QUIET 1)
+  set(orig_CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES})
+  list(LENGTH try_cxx_flags try_max)
+  math(EXPR try_max ${try_max}-1)
```
When `try_cxx_flags` is empty (as it is on OSX due to the failed check above), this ends up setting `try_max` to -1. That, unfortunately, causes the `RANGE` to apparently try to execute...forever, which causes a cmake error when doing the `list(GET)`.

jagerman commented:
Try with the fixes I just pushed.

jagerman commented:
What OS X/compiler/cmake version are you using?  In my Sierra install (and on the travis-ci builds, which I think are El Capitan) it seems the compiler ID is set to Clang, not AppleClang.  (I wonder if this was a cmake change?)

jagerman commented:
Aha, AppleClang was a cmake policy change introduced in 3.0, which we don't hit on the test builds because we have a minimum-2.8.12 instruction.  (I'm guessing you're including this from something with a minimum cmake version declared to be >= 3.0).

rdeits commented:
Yes, that's exactly it. I'm on cmake 3.5.

rdeits commented:
Trying with your update now...

rdeits commented:
Yup, that seems to work. Thanks!

wjakob commented:
I've found it necessary to do a full compile and link to check for LTO reliably across platforms (hence the CHECK_CXX_COMPILER_AND_LINKER_FLAGS macro above). For instance, Gold may not be installed or set up properly, causing linking to fail. The ``check_cxx_compiler_flag`` only does a compile but without linking the resulting binary and hence won't catch this.

wjakob commented:
One more unrelated comment: the loop-based solution in this PR is technically correct, but loops are really hideous in cmake ;). Wouldn't it be simpler to just call a macro which tries a particular set of compiler+linker flags and commits them if successful? Then that macro could be called for all possibilities if applicable (which is really just three: thin LTO, LTO, and MSVC thing)

jagerman commented:
`check_cxx_compiler_flag` definitely appears to be doing both compilation and linking; it internally calls `check_cxx_source_compiles`, which itself documents that `CMAKE_REQUIRED_LIBRARIES = list of libraries to link` needs to be set properly.

It also fails if I throw some invalid garbage into the linker flags but not the cxx flags.

jagerman commented:
> .. and in yet another project, the path for Clang is a bit more advanced: I've found it necessary to inform Gold/Darwin linker where they should store their caches for thin LTO to actually kick in.

Is there an easy way to see whether it is kicking in or not?

dean0x7d commented:
With these changes, setting the flags on MSVC is treated the same as the makefile-based builds (condition on `CMAKE_BUILD_TYPE`), but that doesn't work as intended. Since MSVC is a multi-config build system, all build types are generated at once, regardless of `CMAKE_BUILD_TYPE`. Testing the build with these changes, LTO isn't turned ON/OFF depending on Release/Debug on MSVC. That part should probably be reverted.

jagerman commented:
Ah, I see.  I'll revert the MSVC bits.

dean0x7d commented on tools/pybind11Tools.cmake:
```diff
@@ -36,27 +36,78 @@ function(select_cxx_standard)
+
+  if(U_CMAKE_BUILD_TYPE MATCHES REL AND CMAKE_CXX_COMPILER_ID MATCHES "^GNU|Clang$") # *Not* ^(GNU|Clang)$
+    # Enable link time optimization
+    if(NOT CMAKE_CXX_FLAGS MATCHES "-flto")
```
What's the intent here?

jagerman replied:
I don't remember adding this, but apparently I did!  I'll remove it.

jagerman replied:
Aha, I thought I hadn't added this.  This came from @wjakob 's code.

I'm not sure if we should have it here or not.  At a guess, I'd say the intention would be to allow you to specify something like `-flto=full`, in which case we wouldn't touch the flags.

dean0x7d commented on tools/pybind11Tools.cmake:
```diff
@@ -36,27 +36,78 @@ function(select_cxx_standard)
+# Internal: try compiling with the given c++ and linker flags; if it works, set success_out to 1 and
+# set cxx_out and linker_out in the *caller*'s parent scope and return from the *caller* (this is a
+# macro, not a function!)
+function(_pybind11_check_cxx_and_linker_flags cxxflags linkerflags cxx_out linker_out success_out)
```
Comment:
> this is a macro, not a function!

jagerman replied:
Whoops, fixed.

jagerman commented:
@dean0x7d - I didn't revert them, but rather applied the same generator expression approach everywhere; it simplifies the non-MSVC cmake code a bit doing it that way as well.

jagerman commented:
... but broke g++.  Investigating.

jagerman commented:
Should be fixed now (it was actually the previous commit that broke gcc).

jagerman commented on tools/pybind11Tools.cmake:
```diff
@@ -36,27 +36,78 @@ function(select_cxx_standard)
+  # Release type (upper case for simpler comparisons)
+  string(TOUPPER "${CMAKE_BUILD_TYPE}" U_CMAKE_BUILD_TYPE)
+
+  if(U_CMAKE_BUILD_TYPE MATCHES REL AND CMAKE_CXX_COMPILER_ID MATCHES "^GNU|Clang$") # *Not* ^(GNU|Clang)$
```
It's irrelevant (and gone) in the updated version.  It now never checks it at all, and uses a generator expression to enable the LTO flags (except when doing a debug build).

dean0x7d commented on tools/pybind11Tools.cmake:
```diff
@@ -51,10 +51,12 @@ endfunction()
+    set(cxx ${cxxflags})
+    set(linker ${linkerflags})
+    separate_arguments(cxx)
+    separate_arguments(linker)
```
Wouldn't `separate_arguments(${linkerflags})` accoplish the same thing?

jagerman replied:
Yeah, that's a better idea, I'll change it.

jagerman replied:
(I assume you mean `separate_arguments(${linker_out})`)

jagerman replied:
Actually I think it might not work because of the PARENT_SCOPE.  But I can move the separate_arguments into the find_lto_flags calling code instead (which seems arguably a more appropriate place for it anyway).

dean0x7d replied:
I meant it as written -- to avoid `set(linker ${linkerflags})`.

But actually, why is the transformation between space-separated and semicolon-separated lists even needed? Can't it be kept uniformly one or the other (`list(APPEND ...)`)?

jagerman replied:
`separate_arguments(linkerflags)` didn't work (silently) because this is a macro with a string literal, and so linkerflags apparently isn't a proper variable.

The space-versus-semicolon is a nuisance imposed by cmake: we need space-separated flags when passing multiple flags to check_cxx_compiler_flag (both in the direct argument and in the overridden `CMAKE_REQUIRED_LIBRARIES` variable), but need semicolon-separated when passing the flags to `target_compile_options()`/`target_link_libraries()`.  Otherwise the latter quotes it and passes it as a single argument `"-flto -fno-fat-lto-objects"`, which of course fails.

jagerman commented:
Is there anything still to address for this PR?  It's a bit annoying to have to work around it on other PRs to be able to test build using clang on linux.

jagerman commented:
The recent gcc-7 build failure also started hitting me on debian sid, broke @ghisvail's 2.0.1-2 package (with the fixed cmake scripts), and will start affecting the stretch builds soon: basically it's triggered by -flto and within-function/lambda classes, as a few test scripts do.  As a temporary measure  the just-pushed commit disables -flto under the GCC builds.  I'll submit a follow-up PR to turn it back on once the bug is fixed, but in the meantime this will keep the stretch travis-ci builds working, and will should get the gcc-7/c++17 build working again.

jagerman commented:
Rebased and squashed.  Also added a temporary `-Wno-int-in-bool-context` for gcc (upstream bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79184).  I thought this one had already been applied in a previous PR, but it isn't there, so must have been lost when I split out some of the non-Eigen changes in the Eigen PR.

ghisvail commented:
@jagerman As a temporary measure, is it enough to add `-fno-lto` to `DEB_CXXFLAGS_MAINT_APPEND` in case the fix does not land in gcc-6 on time?

ghisvail commented:
Answering my own question, no it does not work, I get:
```
CMake Error at tools/pybind11Tools.cmake:133 (target_compile_options):
  Error evaluating generator expression:

    $<:-flto>

  Expression did not evaluate to a known generator expression
```

dean0x7d commented:
Is the global `CMAKE_CXX_FLAGS` check completely necessary? It ends up mixing global variables and target-specific state. Modern CMake is trying very hard to move away from global variables. The ability to turn off LTO is already available with the `NO_EXTRAS` option.

Is there any possibility to merge `_pybind11_check_cxx_and_linker_flags` and `_pybind11_return_if_cxx_and_linker_flags_work` into a single macro just to make it easier to follow? Reading the LTO selection is a bit tricky with 3 levels to go through.

jagerman commented:
> Is the global `CMAKE_CXX_FLAGS` check completely necessary? It ends up mixing global variables and target-specific state. Modern CMake is trying very hard to move away from global variables.

Yes.  We aren't setting them, passing them, or encouraging cmake code to use them: we're only respecting if they are set, which is typically going to happen from outside cmake (e.g. invoking `CXXFLAGS="-whatever" cmake ..`).  The perfect case for this is @ghisvail's: he wants to disable -flto globally for a build, because it's broken with his current $CXX.  We're providing no way at all to override the option to a package builder aside from "patch the cmake source" which really isn't reasonable.

> The ability to turn off LTO is already available with the NO_EXTRAS option.

To be blunt, `NO_EXTRAS` seems like a bad solution: it's one option that controls 5 different, largely unrelated things:
- LTO
- `-fvisibility=hidden`
- stripping of binaries
- big object compilation on Windows
- multithreaded compilation on Windows

Moreover it's only available to cmake code, but the main point of respecting CMAKE_CXX_FLAGS is to respect the person invoking cmake, *not* the cmake code calling `pybind11_add_module`.

Those options really should be split up into independent `pybind_add_module` options (e.g. `NO_LTO`, `NO_HIDDEN`, etc.), but that's outside the scope of this PR.  And even if they were split up, it still wouldn't address the issue of the cmake invoker being allowed to override compilation flags with CXXFLAGS.

> Is there any possibility to merge `_pybind11_check_cxx_and_linker_flags` and `_pybind11_return_if_cxx_and_linker_flags_work` into a single macro just to make it easier to follow? Reading the LTO selection is a bit tricky with 3 levels to go through.

It's possible, but it would result in more code: the main point of abstracting the bits that are in the function is that cmake deals with setting up a variable stack so that we don't need to save the local values of `CMAKE_REQUIRED_{QUIET,LIBRARIES}` and restore them afterwards.  Of course this could be done manually inside the macro, but it would increase the macro by more than the size of the function, which I think would end up harder to follow than it is now.  The macro also can't be a function because it really is working as a macro--encapsulating the return and "return value" (i.e. parent scope sets) from the caller's function.

jagerman commented:
@ghisvail - the easiest temporary fix for 2.0.1 on debian would be to apply this patch:

```patch
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 763ad54..870866e 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -65,7 +65,7 @@ if(PYBIND11_TEST_FILES_EIGEN_I GREATER -1)
 endif()
 
 # Create the binding library
-pybind11_add_module(pybind11_tests pybind11_tests.cpp
+pybind11_add_module(pybind11_tests NO_EXTRAS pybind11_tests.cpp
   ${PYBIND11_TEST_FILES} ${PYBIND11_HEADERS})
 
 pybind11_enable_warnings(pybind11_tests)
```

It will bloat up the resulting test .so to about three times the size (about half of the increase because it also disables `-fvisibility=hidden`, the other half because it doesn't strip), but that isn't really a problem here since that's just the test suite, not something you're distributing.

ghisvail commented:
@jagerman I was responding to your bug report on the BTS, asking your for a minimal patch such as this one.

Thanks for the patch. I'll try it right away.

> but that isn't really a problem here since that's just the test suite, not something you're distributing.

You are correct.

dean0x7d commented:
> We aren't setting them, passing them, or encouraging cmake code to use them: we're only respecting if they are set, which is typically going to happen from outside cmake

So is this just a temporary workaround or is it meant to be a standard feature? I just don't see it being viable for the latter. It would be extremely tedious to need to check globals before setting every single flag. Right now it's just LTO on gcc and clang. What about the other compiler and flag combinations (visibility, stripping, ...)? Should `CMAKE_CXX_FLAGS` be the only one checked or do the `CMAKE_CXX_FLAGS_<config>`variants also need to be considered? What about the globals which are specific to the linker or module/shared library flags? Check them all before trying to set any new flag?

> Those options really should be split up into independent pybind_add_module options (e.g. NO_LTO, NO_HIDDEN, etc.)

Sure, individual options can be added as needed. But keep in mind that some fine-grained customization is already possible without any special support. For example:

```cmake
pybind11_add_module(example example.cpp)
# Turn off LTO and restore default visibility
target_compile_options(example PRIVATE -fno-lto -fvisibility=default)
```

The idea for `NO_EXTRAS` was to avoid overcomplicating the function itself. `NO_EXTRAS` is a big button which turns off all non-essentials for users who need to make a fully custom configuration.

jagerman commented:
> ... (hostility) ...

You don't like it, fine; I'm entirely open to alternatives (of which you haven't yet suggested one).  It shouldn't be necessary to patch a CMakeLists.txt file to be able to build the test suite without forcing on LTO flags, which is the main issue.  Allowing "CXXFLAGS=-fno-lto" seemed a very simple approach to me, but clearly it offends you.

jagerman commented:
How about a `PYBIND11_LTO_FLAGS` variable.  If unset, we do the detection as happens now; if specified we use it as is for the LTO flags (could be empty, custom LTO flags, whatever).

jagerman commented:
Only thing I'm unsure of: can I easily distinguish between an option set to empty and an option not specified at all?

jagerman commented:
I updated the PR to store the LTO flags in an internal cache variable, which accomplishes two things: we can skip all the LTO checking for cmake re-runs, and a caller that really wants to disable LTO can override the flags by invoking cmake with `-DPYBIND11_LTO_CXX_FLAGS=`.

dean0x7d commented:
@jagerman I'm sorry for being so critical of the `CXXFLAGS` approach without suggesting any constructive alternatives. If my last post appeared too strong, it's mostly because of some past issues with global flag handling in old-style CMake and Makefile builds. I definitely didn't intend it with any hostility toward you and I'm sorry if it came across that way. (I can get overly opinionated, so feel free to call me out on it.)

jagerman commented:
Thanks @dean0x7d.  I also overreacted a bit, and apologize for that.  I do get your point about both global flags being the wrong place, and the inconsistency of making it work for this flag and not others.

Thoughts on the updated approach?

jagerman commented:
Annoyingly, the updated, broken GCC just migrated to stretch, which means GCC 6 travis-ci builds (outside this PR with the disable-LTO temporary fix) are going to start failing.

dean0x7d commented:
Looks good to me. Moving the target inside the function instead of returning values is also a nice improvement (CMake really makes return values a pain). Minor: The comment above `_pybind11_return_if_cxx_and_linker_flags_work` is outdated -- not a macro and doesn't return anymore.

jagerman commented:
> Minor: The comment above _pybind11_return_if_cxx_and_linker_flags_work is outdated

Good catch; fixed and squashed that in.

wjakob commented on tools/pybind11Tools.cmake:
```diff
@@ -36,61 +36,72 @@ function(select_cxx_standard)
-  message(STATUS "LTO disabled (not supported by the compiler and/or linker)")
+  # Enable LTO flags if found, except for Debug builds
+  if (PYBIND11_LTO_CXX_FLAGS)
+    target_compile_options(${target_name} PRIVATE "$<$<NOT:$<CONFIG:Debug>>:${PYBIND11_LTO_CXX_FLAGS}>")
```
nice!

wjakob commented:
Merged -- apologies that it took me a while to get to this.

jagerman commented:
> Merged -- apologies that it took me a while to get to this.

No worries about the time.

The temporary GCC 6/7 disable-lto workaround should be droppable in a few days--it's fixed upstream in both GCC 6 and 7 branches.  I'll confirm and submit a PR once stretch and/or experimental pick up updated snapshots.

jagerman commented:
@ghisvail - sid has a fixed g++ now, which should be able to build the test suite without issue.

ghisvail commented:
@jagerman fixed with `pybind11/2.0.1-2` in sid

I'll wait that a suitable version of gcc-6 transitions to Stretch and then submit an unblock request. Meanwhile, feel free to verify that this updated package works as intended.
